[Merged by Bors] - feat(Algebra): more instances of subsingleton units

YaelDillies opened the pull request:
Add some missing instances of `Subsingleton Mˣ`.

From Toric


---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

b-mehta reviewed (commented):
To be clear, is the proposal to add `IsSharpMonoid`, even though it's provably equivalent to `Subsingleton Mˣ` in all cases where the former can be written? If this is what you want, it's not clear to me that this is a good idea: let's perhaps discuss this on zulip. Thread: https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/Special-casing.20definitions.20in.20mathlib.3A.20IsSharpMonoid

YaelDillies committed:
feat: sharp monoids
Define sharp monoids, which are the monoids whose only unit is `1`. This concept is useful in log geometry and in particular shows up in the theory of affine monoids.

This concept already existed in the guise of `Subsingleton Mˣ`, but giving it a name makes it easier to refer to and uncovers some missing instances.

From Toric

leanprover-community-bot-assistant commented:
This pull request has conflicts, please merge `master` and resolve them.

vihdzp reviewed (commented):
I'm not really familiar with the mathematics here, but the PR seems sensible and a cursory Google search reveals "sharp monoid" is in fact a term in the literature.

kckennylau commented on Mathlib/Algebra/Group/Units/Basic.lean:
```diff
@@ -287,13 +287,6 @@ instance [Monoid M] : CanLift M Mˣ Units.val IsUnit :=
-section Monoid
-variable [Monoid M]
-
-theorem units_eq_one [Subsingleton Mˣ] (u : Mˣ) : u = 1 := by subsingleton
```
Should we have a deprecation to `Units.eq_one`?

kckennylau commented on Mathlib/Algebra/Group/Units/Defs.lean:
```diff
@@ -425,9 +427,23 @@ lemma IsUnit.exists_left_inv {a : M} (h : IsUnit a) : ∃ b, b * a = 1 := by
+variable (M) in
+/-- A monoid is sharp iff its only unit is `1`. -/
+@[to_additive /-- A monoid is sharp iff its only unit is `0`. -/]
+abbrev IsSharpMonoid := Subsingleton Mˣ
```
I agree. It's redundant because you know that `1` is (syntactically) a unit already.

kckennylau commented on Mathlib/Algebra/Group/Units/Defs.lean:
```diff
@@ -425,9 +427,23 @@ lemma IsUnit.exists_left_inv {a : M} (h : IsUnit a) : ∃ b, b * a = 1 := by
+variable [IsSharpMonoid M]
+
+@[to_additive] lemma Units.eq_one (u : Mˣ) : u = 1 := Subsingleton.elim ..
+@[to_additive] lemma IsUnit.eq_one : IsUnit a → a = 1 := by rintro ⟨u, rfl⟩; simp [u.eq_one]
```
```suggestion
@[to_additive] lemma IsUnit.eq_one (ha : IsUnit a) : a = 1 := congr($ha.unit.eq_one)
```

YaelDillies commented on Mathlib/Algebra/Group/Units/Defs.lean:
```diff
@@ -425,9 +427,23 @@ lemma IsUnit.exists_left_inv {a : M} (h : IsUnit a) : ∃ b, b * a = 1 := by
-  ⟨fun ⟨u, hu⟩ ↦ by rw [← hu, Subsingleton.elim u 1, Units.val_one], fun h ↦ h ▸ isUnit_one⟩
+lemma isUnit_iff_eq_one : IsUnit a ↔ a = 1 where
+  mp := IsUnit.eq_one
+  mpr := by rintro rfl; exact isUnit_one
```
I don't think this is an improvement

YaelDillies committed:
deprecated alias

YaelDillies committed:
drop `IsSharpMonoid`

jcommelin reviewed (commented):
Thanks :tada:

bors merge

b-mehta commented:
Oops, sorry this slipped my mind, and thanks for the merge!
